client/valor/schemas/symbolic/operators.py

```diff
@@ -1,223 +1,234 @@
-            "rhs": self.rhs.to_dict(),
-        }
-
-
-class AppendableFunction(Function):
-    """Base class for defining functions with an unlimited number of arguments."""
-
-    _function = None
-
-    def __init__(self, *args) -> None:
-        """
-        Appendable function.
-        """
-        if len(args) < 2:
-            raise TypeError(
-                f"missing {2 - len(args)} required positional argument"
-            )
-        flat_args = []
-        for arg in args:
-            if isinstance(arg, type(self)):
-                flat_args += arg._args
-            else:
-                flat_args.append(arg)
-        super().__init__(*flat_args)
-
-    def append(self, value: Any):
-        """Appends an argument to the function."""
-        self._args.append(value)
-        return self
+        args = (
+            [arg.to_dict() for arg in self._args]
+            if isinstance(self._args, list)
+            else self._args.to_dict()
+        )
+        return {"op": type(self).__name__.lower(), "args": args}
 
 
-class And(AppendableFunction):
+class And(Function):
     """Implementation of logical AND (&)."""
 
-    _operator = "&"
+    def __init__(self, *args):
+        if len(args) < 2:
+            raise ValueError("Expected at least two arguments.")
+        super().__init__(*args)
 
     def __and__(self, other: Any):
-        self.append(other)
+        if isinstance(other, And):
+            self._args.extend(other._args)
+        else:
+            self._args.append(other)
         return self
 
 
-class Or(AppendableFunction):
+class Or(Function):
     """Implementation of logical OR (|)."""
 
-    _operator = "|"
+    def __init__(self, *args):
+        if len(args) < 2:
+            raise ValueError("Expected at least two arguments.")
+        super().__init__(*args)
 
     def __or__(self, other: Any):
-        self.append(other)
-        return self
-
-
-class Xor(AppendableFunction):
-    """Implementation of logical XOR (^)."""
-
-    _operator = "^"
-
-    def __xor__(self, other: Any):
-        self.append(other)
+        if isinstance(other, Or):
+            self._args.extend(other._args)
+        else:
+            self._args.append(other)
         return self
 
 
-class Negate(OneArgumentFunction):
+class Not(Function):
     """Implementation of logical negation (~)."""
 
-    _operator = "~"
+    def __init__(self, *args):
+        if len(args) != 1:
+            raise ValueError("Negation only takes one argument.")
+        elif isinstance(args[0], Not):
+            return args[0]._args
+        super().__init__(*args)
 
     def __invert__(self):
         """Inverts negation so return contents."""
-        return self.arg
+        if isinstance(self._args, list):
+            raise ValueError("Negation only takes one argument.")
+        return self._args
 
 
-class IsNull(OneArgumentFunction):
+class IsNull(Condition):
     """Implementation of is null value check."""
 
     pass
```

Comment:
should these functions all just pass? I'm confused why we need to define these separately

Reply:
These inherit `Condition` which initializes its op as `self.op = type(self).__name__.lower()`.

This allows client code to rely on a single source of truth (e.g. `Condition`)